cpp/ql/src/experimental/Security/CWE/CWE-190/DangerousUseOfTransformationAfterOperation.ql

```diff
@@ -0,0 +1,94 @@
+ * @description By using the transformation after the operation, you are doing a pointless and dangerous action.
+ * @kind problem
+ * @id cpp/dangerous-use-of-transformation-after-operation.
+ * @problem.severity warning
+ * @precision medium
+ * @tags correctness
+ *       security
+ *       external/cwe/cwe-190
+ */
+
+import cpp
+
+/** Returns the number of the expression in the function call arguments. */
+int argumentPosition(FunctionCall fc, Expr exp, int n) {
+  n in [0 .. fc.getNumberOfArguments() - 1] and
+  fc.getArgument(n) = exp and
+  result = n
+}
+
+/** Holds if a nonsensical type conversion situation is found. */
+predicate conversionDoneLate(MulExpr mexp, Expr e1, Expr e2) {
+  mexp.getConversion().hasExplicitConversion() and
+  mexp.getConversion() instanceof ParenthesisExpr and
+  mexp.getConversion().getConversion() instanceof CStyleCast and
+  mexp.getConversion().getConversion().getType().getSize() > mexp.getType().getSize() and
+  mexp.getConversion().getConversion().getType().getSize() > e2.getType().getSize() and
+  mexp.getConversion().getConversion().getType().getSize() > e1.getType().getSize() and
+  exists(Expr e0 |
+    e0.(AssignExpr).getRValue() = mexp.getParent*() and
+    e0.(AssignExpr).getLValue().getType().getSize() =
+      mexp.getConversion().getConversion().getType().getSize()
+    or
+    mexp.getEnclosingElement().(ComparisonOperation).hasOperands(mexp, e0) and
+    e0.getType().getSize() = mexp.getConversion().getConversion().getType().getSize()
+    or
+    e0.(FunctionCall)
+        .getTarget()
+        .getParameter(argumentPosition(e0.(FunctionCall), mexp, _))
+        .getType()
+        .getSize() = mexp.getConversion().getConversion().getType().getSize()
+  )
+}
+
+/** Holds if the situation of a possible signed overflow used in pointer arithmetic is found. */
+predicate signSmallerWithEqualSizes(MulExpr mexp, Expr e1, Expr e2) {
+  mexp.getConversion+().getUnderlyingType().getSize() = e1.getUnderlyingType().getSize() and
+  (
+    e2.isConstant() or
+    mexp.getConversion+().getUnderlyingType().getSize() = e2.getUnderlyingType().getSize()
+  ) and
+  mexp.getConversion+().getUnderlyingType().getSize() = e1.getUnderlyingType().getSize() and
+  exists(AssignExpr ae |
+    ae.getRValue() = mexp.getParent*() and
+    ae.getRValue().getUnderlyingType().(IntegralType).isUnsigned() and
+    ae.getLValue().getUnderlyingType().(IntegralType).isSigned() and
+    (
+      not exists(DivExpr de | mexp.getParent*() = de)
+      or
+      exists(DivExpr de, Expr ec |
+        e2.isConstant() and
+        de.hasOperands(mexp.getParent*(), ec) and
+        ec.isConstant() and
+        e2.getValue().toInt() > ec.getValue().toInt()
+      )
+    ) and
+    exists(PointerAddExpr pa |
+      ae.getASuccessor+() = pa and
+      pa.getAnOperand().(VariableAccess).getTarget() = ae.getLValue().(VariableAccess).getTarget()
+    )
+  )
+}
+
+from MulExpr mexp, string msg
```

Comment:
Can you add a comment around here that explains why this query catches things that `cpp/integer-multiplication-cast-to-long` doesn't catch?

Reply:
how do you view this text?

```
// Search for an explicit conversion of the result of multiplication in the situation of multiplying types of a smaller size than the result 
// or casting the result of multiplication of comparable types to signed ones.
// In these situations, it is necessary to use the conversion not of the result, but of the arguments.
```

Reply:
That still doesn't explain (at least not to me) what the difference between this query and `cpp/integer-multiplication-cast-to-long` is. As far as I remember, the goal of `cpp/integer-multiplication-cast-to-long` is also to detect cases like:
```cpp
int i1 = ...;
int i2 = ...;
long j = i * i;
```
where there's an implicit cast on the result of the multiplication, but where the cast should really be on the arguments.

Is the difference just that this query requires the cast to be explicit, while `cpp/integer-multiplication-cast-to-long` requires the cast to be implicit?

Reply:
yes, the idea of the request is that the developer uses an explicit override, which would have happened anyway, therefore, this is either stylistically incorrect or overflow.

Reply:
I see. Thanks for clearing that up. Could you extend the description you wrote [here](https://github.com/github/codeql/pull/8247#discussion_r816512144) to say that this query finds cases similar to the ones found by `cpp/integer-multiplication-cast-to-long`, but requires the cast to be explicit?